exchanges/kucoin/kucoin_wrapper.go

```diff
@@ -0,0 +1,1489 @@
+package kucoin
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/config"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/account"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/deposit"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/kline"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/order"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/protocol"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/request"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/stream/buffer"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/ticker"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/trade"
+	"github.com/thrasher-corp/gocryptotrader/log"
+	"github.com/thrasher-corp/gocryptotrader/portfolio/withdraw"
+)
+
+// GetDefaultConfig returns a default exchange config
+func (ku *Kucoin) GetDefaultConfig(ctx context.Context) (*config.Exchange, error) {
+	ku.SetDefaults()
+	exchCfg := new(config.Exchange)
+	exchCfg.Name = ku.Name
+	exchCfg.HTTPTimeout = exchange.DefaultHTTPTimeout
+	exchCfg.BaseCurrencies = ku.BaseCurrencies
+
+	err := ku.SetupDefaults(exchCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if ku.Features.Supports.RESTCapabilities.AutoPairUpdates {
+		err := ku.UpdateTradablePairs(ctx, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return exchCfg, nil
+}
+
+// SetDefaults sets the basic defaults for Kucoin
+func (ku *Kucoin) SetDefaults() {
+	ku.Name = "Kucoin"
+	ku.Enabled = true
+	ku.Verbose = false
+
+	ku.API.CredentialsValidator.RequiresKey = true
+	ku.API.CredentialsValidator.RequiresSecret = true
+	ku.API.CredentialsValidator.RequiresClientID = true
+
+	spot := currency.PairStore{
+		RequestFormat: &currency.PairFormat{Uppercase: true, Delimiter: currency.DashDelimiter},
+		ConfigFormat:  &currency.PairFormat{Uppercase: true, Delimiter: currency.DashDelimiter},
+	}
+	futures := currency.PairStore{
+		RequestFormat: &currency.PairFormat{Uppercase: true},
+		ConfigFormat:  &currency.PairFormat{Uppercase: true, Delimiter: currency.UnderscoreDelimiter},
+	}
+	err := ku.StoreAssetPairFormat(asset.Spot, spot)
+	if err != nil {
+		log.Errorln(log.ExchangeSys, err)
+	}
+	err = ku.StoreAssetPairFormat(asset.Margin, spot)
+	if err != nil {
+		log.Errorln(log.ExchangeSys, err)
+	}
+	err = ku.StoreAssetPairFormat(asset.Futures, futures)
+	if err != nil {
+		log.Errorln(log.ExchangeSys, err)
+	}
+	ku.Features = exchange.Features{
+		Supports: exchange.FeaturesSupported{
+			REST:      true,
+			Websocket: true,
+			RESTCapabilities: protocol.Features{
+				TickerFetching:    true,
```

Comment:
I've missed commenting on this in your other PRs, its very important
Please add `TickerBatching: true,`